File: src/csharp/Foundgine.Core/Semantic/Planning/SemanticEquivalenceFingerprint.cs
using System.Globalization;
using System.Text;
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Query;

namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Produces a canonical semantic identity used to determine whether two plans
/// have the same provider-neutral meaning. Unlike the execution fingerprint,
/// this representation normalizes only transformations that Foundgine defines
/// as semantically equivalent.
/// </summary>
public static class SemanticEquivalenceFingerprint
{
    public static string Create(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        var builder = new StringBuilder(512);
        builder.Append("semantic-v1|");
        AppendSecurity(builder, plan.EffectiveSecurityInvariants);
        AppendNode(builder, plan.Root);
        return builder.ToString();
    }

    private static void AppendSecurity(StringBuilder builder, IReadOnlyList<string> invariants)
    {
        builder.Append("security[");
        foreach (var invariant in invariants.OrderBy(x => x, StringComparer.Ordinal))
            builder.Append(invariant).Append(',');
        builder.Append("]|");
    }

    private static void AppendNode(StringBuilder builder, SemanticPlanNode node)
    {
        builder.Append("node(").Append(node.Id).Append('|')
            .Append((byte)node.Operation).Append('|')
            .Append(node.EntityId.Value).Append('|')
            .Append(node.ViaRelationship?.Value.ToString() ?? "-").Append('|')
            .Append(node.ViaConnection?.Value.ToString() ?? "-").Append(')');

        builder.Append("fields[");
        var seenFields = new HashSet<ulong>();
        foreach (var field in node.Fields)
        {
            if (seenFields.Add(field.Value))
                builder.Append(field.Value).Append(',');
        }

        builder.Append("]|");

        AppendQueryOptions(builder, node.QueryOptions);
        builder.Append("auth[");
        AppendAuthorization(builder, node.Authorization);
        builder.Append("]|");

        builder.Append("children[");
        foreach (var child in node.Children)
            AppendNode(builder, child);
        builder.Append(']');
    }

    private static void AppendQueryOptions(StringBuilder builder, SemanticQueryOptions? options)
    {
        if (options is null)
        {
            builder.Append("query[-]|");
            return;
        }

        builder.Append("query[")
            .Append(options.Limit?.ToString(CultureInfo.InvariantCulture) ?? "-").Append('|')
            .Append(options.Offset?.ToString(CultureInfo.InvariantCulture) ?? "-").Append('|');
        AppendValue(builder, options.After);
        builder.Append("|order[");
        foreach (var term in options.EffectiveOrder)
        {
            builder.Append(term.Field.Value).Append(':')
                .Append((byte)term.Direction).Append(':')
                .Append((byte)term.Aggregate).Append(':');
            foreach (var relationship in term.EffectivePath)
                builder.Append(relationship.Value).Append('.');
            builder.Append(',');
        }

        builder.Append("]|");
        AppendFilter(builder, options.Filter);
        builder.Append("]|");
    }

    private static void AppendFilter(StringBuilder builder, SemanticFilterExpression? filter)
    {
        if (filter is not null)
            filter = CanonicalizeAggregateRelationshipPushdown(filter);

        switch (filter)
        {
            case null:
                builder.Append("-");
                return;
            case SemanticFieldFilter field:
                builder.Append("field(").Append(field.Field.Value).Append('|')
                    .Append((byte)field.Operator).Append('|');
                AppendValue(builder, field.Value);
                builder.Append(')');
                return;
            case SemanticRelationshipFilter relationship:
                builder.Append("relationship(").Append(relationship.Relationship.Value).Append('|')
                    .Append((byte)relationship.Quantifier).Append('|');
                AppendFilter(builder, relationship.Predicate);
                builder.Append(')');
                return;
            case SemanticAggregateFilter aggregate:
                builder.Append("aggregate(").Append(aggregate.Relationship.Value).Append('|')
                    .Append((byte)aggregate.Aggregate).Append('|')
                    .Append(aggregate.Field?.Value.ToString() ?? "-").Append('|')
                    .Append((byte)aggregate.Operator).Append('|');
                AppendValue(builder, aggregate.Value);
                builder.Append('|');
                AppendFilter(builder, aggregate.Predicate);
                builder.Append(')');
                return;
            case SemanticAndFilter and:
                AppendCanonicalBooleanFilter(builder, and);
                return;
            case SemanticOrFilter or:
                AppendCanonicalBooleanFilter(builder, or);
                return;
            default:
                throw new NotSupportedException(
                    $"Cannot establish semantic equivalence for filter '{filter.GetType().Name}'.");
        }
    }

    private static void AppendCanonicalBooleanFilter(StringBuilder builder, SemanticFilterExpression filter)
    {
        // AggregateRelationshipFilterPushdownRule replaces:
        //   COUNT(R) > 0 AND SOME(R, P)
        // with:
        //   COUNT(R WHERE P) > 0
        // This is a semantic normalization, not an execution-plan identity. Keep it
        // here (rather than SemanticPlanFingerprint) so the rewrite composer can prove
        // the transformation without weakening exact plan/cache fingerprints.
        filter = CanonicalizeAggregateRelationshipPushdown(filter);

        const int maxTerms = 32;
        var terms = ToDnf(filter, maxTerms);
        if (terms is null)
        {
            switch (filter)
            {
                case SemanticAndFilter and:
                    AppendCommutativeFilter(builder, "and", and.Expressions);
                    return;
                case SemanticOrFilter or:
                    AppendCommutativeFilter(builder, "or", or.Expressions);
                    return;
                default:
                    throw new InvalidOperationException("Expected a boolean filter.");
            }
        }

        builder.Append("dnf[");
        foreach (var term in terms
                     .Select(t => t.OrderBy(CreateFilterKey, StringComparer.Ordinal).Select(CreateFilterKey).ToArray())
                     .OrderBy(t => string.Join("&", t), StringComparer.Ordinal))
        {
            builder.Append('(');
            foreach (var atom in term)
                builder.Append(atom).Append(',');
            builder.Append(')');
        }

        builder.Append(']');
    }

    private static SemanticFilterExpression CanonicalizeAggregateRelationshipPushdown(
        SemanticFilterExpression filter)
    {
        // Canonicalize the logical equivalence in the correct order. For an AND
        // containing COUNT(R) > 0 and SOME(R, P), the pair must be recognized
        // before the COUNT expression is individually normalized to SOME(R, null).
        // Otherwise the two operands become indistinguishable relationship filters
        // and the semantic fingerprint cannot prove the pushdown.
        switch (filter)
        {
            case SemanticAndFilter and:
            {
                var expressions = and.Expressions
                    .Select(CanonicalizeAggregateRelationshipPushdown)
                    .ToList();

                // The recursive normalization above may already have converted a
                // filtered COUNT into SOME. Therefore perform the equivalence in a
                // second canonical pass over the original AND shape when possible.
                var raw = and.Expressions.ToList();
                for (var i = 0; i < raw.Count; i++)
                {
                    if (raw[i] is not SemanticAggregateFilter aggregate ||
                        aggregate.Aggregate != SemanticFilterAggregate.Count ||
                        aggregate.Field is not null ||
                        aggregate.Predicate is not null ||
                        !IsCountExistenceComparison(aggregate))
                        continue;

                    for (var j = 0; j < raw.Count; j++)
                    {
                        if (i == j || raw[j] is not SemanticRelationshipFilter relationship ||
                            relationship.Quantifier != SemanticRelationshipQuantifier.Some ||
                            relationship.Relationship != aggregate.Relationship)
                            continue;

                        var pushed = aggregate with { Predicate = relationship.Predicate };
                        var remaining = new List<SemanticFilterExpression>();
                        for (var k = 0; k < raw.Count; k++)
                        {
                            if (k == i) remaining.Add(pushed);
                            else if (k != j) remaining.Add(raw[k]);
                        }

                        var merged = remaining.Count switch
                        {
                            0 => throw new InvalidOperationException(
                                "Aggregate equivalence normalization produced an empty AND expression."),
                            1 => remaining[0],
                            _ => new SemanticAndFilter(remaining)
                        };

                        return CanonicalizeAggregateRelationshipPushdown(merged);
                    }
                }

                return expressions.Count switch
                {
                    0 => throw new InvalidOperationException(
                        "Aggregate equivalence normalization produced an empty AND expression."),
                    1 => expressions[0],
                    _ => new SemanticAndFilter(expressions)
                };
            }

            case SemanticAggregateFilter aggregate when
                aggregate.Aggregate == SemanticFilterAggregate.Count &&
                aggregate.Field is null &&
                aggregate.Predicate is not null:
            {
                var strategy = AggregateExecutionStrategyResolver.Resolve(
                    aggregate.Operator, aggregate.Value);
                if (strategy is AggregateExecutionStrategy.CountExistsShortCircuit or
                    AggregateExecutionStrategy.CountEmptyShortCircuit)
                {
                    return new SemanticRelationshipFilter(
                        aggregate.Relationship,
                        strategy == AggregateExecutionStrategy.CountEmptyShortCircuit
                            ? SemanticRelationshipQuantifier.None
                            : SemanticRelationshipQuantifier.Some,
                        CanonicalizeAggregateRelationshipPushdown(aggregate.Predicate));
                }

                return aggregate with
                {
                    Predicate = CanonicalizeAggregateRelationshipPushdown(aggregate.Predicate)
                };
            }

            case SemanticOrFilter or:
            {
                var expressions = or.Expressions
                    .Select(CanonicalizeAggregateRelationshipPushdown)
                    .ToArray();
                return new SemanticOrFilter(expressions);
            }

            case SemanticRelationshipFilter relationship:
            {
                var predicate = CanonicalizeAggregateRelationshipPushdown(relationship.Predicate);
                return ReferenceEquals(predicate, relationship.Predicate)
                    ? relationship
                    : relationship with { Predicate = predicate };
            }

            default:
                return filter;
        }
    }

    private static bool IsCountExistenceComparison(SemanticAggregateFilter aggregate)
    {
        if (!TryGetIntegral(aggregate.Value, out var value))
            return false;

        return aggregate.Operator switch
        {
            SemanticAggregateFilterOperator.Gt when value == 0 => true,
            SemanticAggregateFilterOperator.Gte when value == 1 => true,
            SemanticAggregateFilterOperator.Neq when value == 0 => true,
            _ => false
        };
    }

    private static bool TryGetIntegral(object? value, out long result)
    {
        switch (value)
        {
            case byte v:
                result = v;
                return true;
            case sbyte v:
                result = v;
                return true;
            case short v:
                result = v;
                return true;
            case ushort v:
                result = v;
                return true;
            case int v:
                result = v;
                return true;
            case uint v:
                result = v;
                return true;
            case ulong v when v <= long.MaxValue:
                result = (long)v;
                return true;
            case long v:
                result = v;
                return true;
            default:
                result = 0;
                return false;
        }
    }

    private static IReadOnlyList<IReadOnlyList<SemanticFilterExpression>>? ToDnf(
        SemanticFilterExpression expression,
        int maxTerms)
    {
        switch (expression)
        {
            case SemanticAndFilter and:
            {
                IReadOnlyList<SemanticFilterExpression> emptyTerm = Array.Empty<SemanticFilterExpression>();
                var result = new List<IReadOnlyList<SemanticFilterExpression>> { emptyTerm };
                foreach (var child in and.Expressions)
                {
                    var childTerms = ToDnf(child, maxTerms);
                    if (childTerms is null) return null;
                    var next = new List<IReadOnlyList<SemanticFilterExpression>>();
                    foreach (var left in result)
                    foreach (var right in childTerms)
                    {
                        var combined = left.Concat(right).ToArray();
                        next.Add(combined);
                        if (next.Count > maxTerms) return null;
                    }

                    result = next;
                }

                return result;
            }
            case SemanticOrFilter or:
            {
                var result = new List<IReadOnlyList<SemanticFilterExpression>>();
                foreach (var child in or.Expressions)
                {
                    var childTerms = ToDnf(child, maxTerms);
                    if (childTerms is null) return null;
                    result.AddRange(childTerms);
                    if (result.Count > maxTerms) return null;
                }

                return result;
            }
            default:
                return [[expression]];
        }
    }

    private static void AppendCommutativeFilter(
        StringBuilder builder,
        string kind,
        IReadOnlyList<SemanticFilterExpression> expressions)
    {
        var keys = expressions.Select(CreateFilterKey)
            .OrderBy(x => x, StringComparer.Ordinal)
            .ToArray();
        builder.Append(kind).Append('[');
        foreach (var key in keys)
            builder.Append(key).Append(',');
        builder.Append(']');
    }

    private static string CreateFilterKey(SemanticFilterExpression expression)
    {
        var builder = new StringBuilder(128);
        AppendFilter(builder, expression);
        return builder.ToString();
    }

    private static void AppendAuthorization(StringBuilder builder, AuthorizationPredicate? predicate)
    {
        if (predicate is null)
            return;

        var canonical = CanonicalAuthorization(predicate);
        AppendAuthorizationNode(builder, canonical);
    }

    private static AuthorizationPredicate CanonicalAuthorization(AuthorizationPredicate predicate)
    {
        var left = predicate.Left is null ? null : CanonicalAuthorization(predicate.Left);
        var right = predicate.Right is null ? null : CanonicalAuthorization(predicate.Right);
        var current = predicate with { Left = left, Right = right };

        if (current.Kind == AuthorizationPredicateKind.Not &&
            current.Left?.Kind == AuthorizationPredicateKind.Not &&
            current.Left.Left is not null)
            return CanonicalAuthorization(current.Left.Left);

        if (current.Kind is AuthorizationPredicateKind.And or AuthorizationPredicateKind.Or)
        {
            var operands = new List<AuthorizationPredicate>();
            FlattenAuthorization(current.Kind, current, operands);
            operands = operands
                .Select(CanonicalAuthorization)
                .GroupBy(CreateAuthorizationKey, StringComparer.Ordinal)
                .Select(group => group.First())
                .OrderBy(CreateAuthorizationKey, StringComparer.Ordinal)
                .ToList();

            if (operands.Count == 1)
                return operands[0];

            AuthorizationPredicate result = operands[0];
            for (var i = 1; i < operands.Count; i++)
                result = current.Kind == AuthorizationPredicateKind.And
                    ? AuthorizationPredicate.And(result, operands[i])
                    : AuthorizationPredicate.Or(result, operands[i]);
            return result;
        }

        return current;
    }

    private static void FlattenAuthorization(
        AuthorizationPredicateKind kind,
        AuthorizationPredicate node,
        ICollection<AuthorizationPredicate> destination)
    {
        if (node.Kind == kind)
        {
            if (node.Left is not null)
                FlattenAuthorization(kind, node.Left, destination);
            if (node.Right is not null)
                FlattenAuthorization(kind, node.Right, destination);
            return;
        }

        destination.Add(node);
    }

    private static string CreateAuthorizationKey(AuthorizationPredicate predicate)
    {
        var builder = new StringBuilder(96);
        AppendAuthorizationNode(builder, predicate);
        return builder.ToString();
    }

    private static void AppendAuthorizationNode(StringBuilder builder, AuthorizationPredicate predicate)
    {
        builder.Append((byte)predicate.Kind).Append('(');
        AppendValue(builder, predicate.Name);
        builder.Append('|');
        AppendValue(builder, predicate.Value);
        builder.Append('|');
        if (predicate.Left is not null)
            AppendAuthorizationNode(builder, predicate.Left);
        builder.Append('|');
        if (predicate.Right is not null)
            AppendAuthorizationNode(builder, predicate.Right);
        builder.Append(')');
    }

    private static void AppendValue(StringBuilder builder, object? value)
    {
        if (value is null)
        {
            builder.Append("null");
            return;
        }

        var type = value.GetType();
        builder.Append(type.AssemblyQualifiedName).Append('=');
        switch (value)
        {
            case string text:
                builder.Append(text.Length).Append(':').Append(text);
                break;
            case byte[] bytes:
                builder.Append(Convert.ToHexString(bytes));
                break;
            case System.Collections.IEnumerable sequence when value is not string:
                builder.Append('[');
                foreach (var item in sequence)
                {
                    AppendValue(builder, item);
                    builder.Append(',');
                }

                builder.Append(']');
                break;
            case IFormattable formattable:
                builder.Append(formattable.ToString(null, CultureInfo.InvariantCulture));
                break;
            default:
                builder.Append(value);
                break;
        }
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/SemanticEquivalenceFingerprint.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.abstractions.AuthorizationPredicate;
import com.foundgine.core.abstractions.AuthorizationPredicateKind;
import com.foundgine.core.semantic.query.*;
import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * Produces a canonical semantic identity used to determine whether two plans
 * have the same provider-neutral meaning. Unlike the execution fingerprint,
 * this representation normalizes only transformations defined as semantically
 * equivalent.
 */
public final class SemanticEquivalenceFingerprint {
	private SemanticEquivalenceFingerprint() {
	}

	public static String create(SemanticPlan plan) {
		Objects.requireNonNull(plan, "plan");
		StringBuilder b = new StringBuilder(512);
		b.append("semantic-v1|");
		b.append("security[");
		plan.effectiveSecurityInvariants().stream().sorted().forEach(x -> b.append(x).append(','));
		b.append("]|");
		appendNode(b, plan.root());
		return b.toString();
	}

	private static void appendNode(StringBuilder b, SemanticPlanNode node) {
		b.append("node(").append(node.id()).append('|').append((byte) node.operation().ordinal()).append('|')
				.append(node.entityId().value()).append('|')
				.append(node.viaRelationship() != null ? node.viaRelationship().value() : "-").append('|')
				.append(node.viaConnection() != null ? node.viaConnection().value() : "-").append(')');
		b.append("fields[");
		Set<Long> seen = new HashSet<>();
		for (var f : node.fields())
			if (seen.add(f.value()))
				b.append(f.value()).append(',');
		b.append("]|");
		appendQueryOptions(b, node.queryOptions());
		b.append("auth[");
		appendAuthorization(b, node.authorization());
		b.append("]|");
		b.append("children[");
		node.children().forEach(c -> appendNode(b, c));
		b.append(']');
	}

	private static void appendQueryOptions(StringBuilder b, SemanticQueryOptions options) {
		if (options == null) {
			b.append("query[-]|");
			return;
		}
		b.append("query[").append(options.limit() != null ? options.limit() : "-").append('|')
				.append(options.offset() != null ? options.offset() : "-").append('|');
		appendValue(b, options.after());
		b.append("|order[");
		for (var term : options.effectiveOrder()) {
			b.append(term.field().value()).append(':').append((byte) term.direction().ordinal()).append(':')
					.append((byte) term.aggregate().ordinal()).append(':');
			for (var r : term.effectivePath())
				b.append(r.value()).append('.');
			b.append(',');
		}
		b.append("]|");
		appendFilter(b, options.filter());
		b.append("]|");
	}

	private static void appendFilter(StringBuilder b, SemanticFilterExpression filter) {
		filter = canonicalizeAggregateRelationshipPushdown(filter);
		if (filter == null) {
			b.append('-');
			return;
		}
		if (filter instanceof SemanticFieldFilter f) {
			b.append("field(").append(f.field().value()).append('|').append((byte) f.operator().ordinal()).append('|');
			appendValue(b, f.value());
			b.append(')');
			return;
		}
		if (filter instanceof SemanticRelationshipFilter r) {
			b.append("relationship(").append(r.relationship().value()).append('|')
					.append((byte) r.quantifier().ordinal()).append('|');
			appendFilter(b, r.predicate());
			b.append(')');
			return;
		}
		if (filter instanceof SemanticAggregateFilter a) {
			b.append("aggregate(").append(a.relationship().value()).append('|').append((byte) a.aggregate().ordinal())
					.append('|').append(a.field() != null ? a.field().value() : "-").append('|')
					.append((byte) a.operator().ordinal()).append('|');
			appendValue(b, a.value());
			b.append('|');
			appendFilter(b, a.predicate());
			b.append(')');
			return;
		}
		if (filter instanceof SemanticAndFilter a) {
			appendCanonicalBooleanFilter(b, a);
			return;
		}
		if (filter instanceof SemanticOrFilter o) {
			appendCanonicalBooleanFilter(b, o);
			return;
		}
		throw new UnsupportedOperationException(
				"Cannot establish semantic equivalence for filter '" + filter.getClass().getSimpleName() + "'.");
	}

	private static void appendCanonicalBooleanFilter(StringBuilder b, SemanticFilterExpression filter) {
		var terms = toDnf(filter, 32);
		if (terms == null) {
			if (filter instanceof SemanticAndFilter a)
				appendCommutativeFilter(b, "and", a.expressions());
			else if (filter instanceof SemanticOrFilter o)
				appendCommutativeFilter(b, "or", o.expressions());
			else
				throw new IllegalStateException("Expected boolean filter.");
			return;
		}
		b.append("dnf[");
		var normalized = new ArrayList<List<String>>();
		for (var term : terms) {
			var keys = term.stream().map(SemanticEquivalenceFingerprint::createFilterKey).sorted().toList();
			normalized.add(keys);
		}
		normalized.sort(Comparator.comparing(x -> String.join("&", x)));
		for (var term : normalized) {
			b.append('(');
			term.forEach(x -> b.append(x).append(','));
			b.append(')');
		}
		b.append(']');
	}

	private static SemanticFilterExpression canonicalizeAggregateRelationshipPushdown(SemanticFilterExpression filter) {
		if (filter == null)
			return null;
		if (filter instanceof SemanticAndFilter a) {
			var raw = a.expressions();
			for (int i = 0; i < raw.size(); i++) {
				if (!(raw.get(i) instanceof SemanticAggregateFilter aggregate)
						|| aggregate.aggregate() != SemanticFilterAggregate.COUNT || aggregate.field() != null
						|| aggregate.predicate() != null || !isCountExistenceComparison(aggregate))
					continue;
				for (int j = 0; j < raw.size(); j++) {
					if (i == j || !(raw.get(j) instanceof SemanticRelationshipFilter r)
							|| r.quantifier() != SemanticRelationshipQuantifier.SOME
							|| !r.relationship().equals(aggregate.relationship()))
						continue;
					var pushed = new SemanticAggregateFilter(aggregate.relationship(), aggregate.aggregate(),
							aggregate.field(), aggregate.operator(), aggregate.value(), r.predicate());
					var remaining = new ArrayList<SemanticFilterExpression>();
					for (int k = 0; k < raw.size(); k++) {
						if (k == i)
							remaining.add(pushed);
						else if (k != j)
							remaining.add(raw.get(k));
					}
					SemanticFilterExpression merged = remaining.size() == 1 ? remaining.get(0)
							: new SemanticAndFilter(remaining);
					return canonicalizeAggregateRelationshipPushdown(merged);
				}
			}
			var expressions = a.expressions().stream()
					.map(SemanticEquivalenceFingerprint::canonicalizeAggregateRelationshipPushdown).toList();
			return expressions.size() == 1 ? expressions.get(0) : new SemanticAndFilter(expressions);
		}
		if (filter instanceof SemanticAggregateFilter a && a.aggregate() == SemanticFilterAggregate.COUNT
				&& a.field() == null && a.predicate() != null) {
			var strategy = AggregateExecutionStrategyResolver.resolve(a.operator(), a.value());
			if (strategy == AggregateExecutionStrategy.COUNT_EXISTS_SHORT_CIRCUIT
					|| strategy == AggregateExecutionStrategy.COUNT_EMPTY_SHORT_CIRCUIT)
				return new SemanticRelationshipFilter(a.relationship(),
						strategy == AggregateExecutionStrategy.COUNT_EMPTY_SHORT_CIRCUIT
								? SemanticRelationshipQuantifier.NONE
								: SemanticRelationshipQuantifier.SOME,
						canonicalizeAggregateRelationshipPushdown(a.predicate()));
			return new SemanticAggregateFilter(a.relationship(), a.aggregate(), a.field(), a.operator(), a.value(),
					canonicalizeAggregateRelationshipPushdown(a.predicate()));
		}
		if (filter instanceof SemanticOrFilter o)
			return new SemanticOrFilter(o.expressions().stream()
					.map(SemanticEquivalenceFingerprint::canonicalizeAggregateRelationshipPushdown).toList());
		if (filter instanceof SemanticRelationshipFilter r) {
			var p = canonicalizeAggregateRelationshipPushdown(r.predicate());
			return p == r.predicate() ? r : new SemanticRelationshipFilter(r.relationship(), r.quantifier(), p);
		}
		return filter;
	}

	private static boolean isCountExistenceComparison(SemanticAggregateFilter a) {
		Long v = tryGetIntegral(a.value());
		if (v == null)
			return false;
		return (a.operator() == SemanticAggregateFilterOperator.GT && v == 0)
				|| (a.operator() == SemanticAggregateFilterOperator.GTE && v == 1)
				|| (a.operator() == SemanticAggregateFilterOperator.NEQ && v == 0);
	}

	private static Long tryGetIntegral(Object value) {
		if (value instanceof Byte v)
			return v.longValue();
		if (value instanceof Short v)
			return v.longValue();
		if (value instanceof Integer v)
			return v.longValue();
		if (value instanceof Long v)
			return v;
		return null;
	}

	private static List<List<SemanticFilterExpression>> toDnf(SemanticFilterExpression e, int max) {
		if (e instanceof SemanticAndFilter a) {
			List<List<SemanticFilterExpression>> result = new ArrayList<>();
			result.add(List.of());
			for (var child : a.expressions()) {
				var childTerms = toDnf(child, max);
				if (childTerms == null)
					return null;
				var next = new ArrayList<List<SemanticFilterExpression>>();
				for (var l : result)
					for (var r : childTerms) {
						var c = new ArrayList<SemanticFilterExpression>(l);
						c.addAll(r);
						next.add(c);
						if (next.size() > max)
							return null;
					}
				result = next;
			}
			return result;
		}
		if (e instanceof SemanticOrFilter o) {
			var result = new ArrayList<List<SemanticFilterExpression>>();
			for (var child : o.expressions()) {
				var t = toDnf(child, max);
				if (t == null)
					return null;
				result.addAll(t);
				if (result.size() > max)
					return null;
			}
			return result;
		}
		return List.of(List.of(e));
	}

	private static void appendCommutativeFilter(StringBuilder b, String kind,
			List<SemanticFilterExpression> expressions) {
		b.append(kind).append('[');
		expressions.stream().map(SemanticEquivalenceFingerprint::createFilterKey).sorted()
				.forEach(x -> b.append(x).append(','));
		b.append(']');
	}

	private static String createFilterKey(SemanticFilterExpression e) {
		StringBuilder b = new StringBuilder();
		appendFilter(b, e);
		return b.toString();
	}

	private static void appendAuthorization(StringBuilder b, AuthorizationPredicate p) {
		if (p != null)
			appendAuthorizationNode(b, canonicalAuthorization(p));
	}

	private static AuthorizationPredicate canonicalAuthorization(AuthorizationPredicate p) {
		var left = p.left() == null ? null : canonicalAuthorization(p.left());
		var right = p.right() == null ? null : canonicalAuthorization(p.right());
		var current = (left == p.left() && right == p.right()) ? p
				: new AuthorizationPredicate(p.kind(), p.name(), p.value(), left, right);
		if (current.kind() == AuthorizationPredicateKind.NOT && current.left() != null
				&& current.left().kind() == AuthorizationPredicateKind.NOT && current.left().left() != null)
			return canonicalAuthorization(current.left().left());
		if (current.kind() == AuthorizationPredicateKind.AND || current.kind() == AuthorizationPredicateKind.OR) {
			var ops = new ArrayList<AuthorizationPredicate>();
			flattenAuthorization(current.kind(), current, ops);
			var unique = new LinkedHashMap<String, AuthorizationPredicate>();
			for (var op : ops) {
				var c = canonicalAuthorization(op);
				unique.putIfAbsent(createAuthorizationKey(c), c);
			}
			ops = new ArrayList<>(unique.values());
			ops.sort(Comparator.comparing(SemanticEquivalenceFingerprint::createAuthorizationKey));
			AuthorizationPredicate result = ops.get(0);
			for (int i = 1; i < ops.size(); i++)
				result = current.kind() == AuthorizationPredicateKind.AND
						? AuthorizationPredicate.and(result, ops.get(i))
						: AuthorizationPredicate.or(result, ops.get(i));
			return result;
		}
		return current;
	}

	private static void flattenAuthorization(AuthorizationPredicateKind kind, AuthorizationPredicate n,
			List<AuthorizationPredicate> out) {
		if (n.kind() == kind) {
			if (n.left() != null)
				flattenAuthorization(kind, n.left(), out);
			if (n.right() != null)
				flattenAuthorization(kind, n.right(), out);
		} else
			out.add(n);
	}

	private static String createAuthorizationKey(AuthorizationPredicate p) {
		StringBuilder b = new StringBuilder();
		appendAuthorizationNode(b, p);
		return b.toString();
	}

	private static void appendAuthorizationNode(StringBuilder b, AuthorizationPredicate p) {
		b.append((byte) p.kind().ordinal()).append('(');
		appendValue(b, p.name());
		b.append('|');
		appendValue(b, p.value());
		b.append('|');
		if (p.left() != null)
			appendAuthorizationNode(b, p.left());
		b.append('|');
		if (p.right() != null)
			appendAuthorizationNode(b, p.right());
		b.append(')');
	}

	private static void appendValue(StringBuilder b, Object value) {
		if (value == null) {
			b.append("null");
			return;
		}
		b.append(value.getClass().getName()).append('=');
		if (value instanceof String s)
			b.append(s.length()).append(':').append(s);
		else if (value instanceof byte[] bytes) {
			for (byte x : bytes)
				b.append(String.format("%02X", x));
		} else if (value instanceof Iterable<?> it) {
			b.append('[');
			for (var x : it) {
				appendValue(b, x);
				b.append(',');
			}
			b.append(']');
		} else
			b.append(value);
	}
}
